File: dynamic_programming/removing_sets.py
n = int(input())

# state : subtract one of the digits to give minimal number of steps to reduce the number to 0
# transition : dp[i] = min(dp of each digits)

dp = [float('inf')] * (n + 1)

dp[0] = 0

for i in range(1,n + 1):
    temp = i

    while temp > 0:
        digit = temp % 10
        
        if digit != 0:
            dp[i] = min(dp[i],dp[i - digit] + 1) 

        temp = temp // 10

print(dp[n])

File: dynamic_programming/removing_sets.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    int n;
    cin >> n;

    vector<int> dp(n + 1, INT_MAX);
    dp[0] = 0;

    for (int i = 1; i <= n; i++) {
        int temp = i;

        while (temp > 0) {
            int digit = temp % 10;

            if (digit != 0 && dp[i - digit] != INT_MAX) {
                dp[i] = min(dp[i], dp[i - digit] + 1);
            }

            temp /= 10;
        }
    }

    cout << dp[n] << "\n";
    return 0;
}